### PR TITLE
Account settings

### PR DESCRIPTION
**Pull request by ordass00**
Added account settings

### CODE REVIEW

File: change_settings/change_settings_functionality.php [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+$request = file_get_contents("php://input");
+$conn = connect_local();
+
+if (isset($request) && !empty($request)) {
+    $reqObj = json_decode($request);
+    switch ($reqObj->method) {
+        case "change_user_information":
+            $updated_user_information = change_user_information($conn, $reqObj->userid, $reqObj->firstname, $reqObj->lastname, $reqObj->username, $reqObj->email, $reqObj->dateOfBirth);
+            if (!$updated_user_information) {
+                echo json_encode(["error" => true, "errorText" => "Failed to change personal information"]);
```
Comment (Zylesto) [inline change_settings/change_settings_functionality.php:15]: Failed to change personal information.

File: change_settings/change_settings_functionality.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            $password_repeat = $reqObj->password_repeat;
+            if ((!empty($password) && !empty($password_repeat))) {
+                if ($password == $password_repeat) {
+                    if (preg_match($pattern, $password) == 1 && preg_match($pattern, $password_repeat) == 1) {
+                        $password_encrypted = password_hash($reqObj->password, PASSWORD_BCRYPT);
+                        $success = reset_password($conn, $_SESSION["userid"], $password_encrypted);
+                        if ($success) {
+                            echo json_encode(["error" => false, "errorText" => ""]);
+                        } else {
+                            echo json_encode(["error" => true, "errorText" => "Could not reset password. Try again later"]);
```
Comment (Zylesto) [inline change_settings/change_settings_functionality.php:40]: All error messages end with an dot

File: change_settings/change_settings.php [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+        window.changeAccountSettings = changeAccountSettings;
+        window.changePassword = changePassword;
+        window.changePreferences = changePreferences;
+        window.changeIngredients = changeIngredients;
+        window.pwdValidation = pwdValidation;
+        window.showHidePassword = showHidePassword;
+        window.validateInput = validateInput;
+    </script>
+
+    <title>Account settings</title>
```
Comment (Zylesto) [inline change_settings/change_settings.php:28]: Acoount Settings

File: change_settings/change_settings.php [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+                    aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
+                <span class="navbar-toggler-icon"></span>
+            </button>
+            <div class="collapse navbar-collapse" id="navbarSupportedContent">
+                <ul class="navbar-nav me-auto mb-2 mb-lg-0">
+                    <li class="nav-item">
+                        <a class="nav-link active" href="../index/index.html">Home</a>
+                    </li>
+                    <li class="nav-item">
+                        <a class="nav-link" href="../meal_plan_overview/meal_plan_overview.php">Meal plan</a>
```
Comment (Zylesto) [inline change_settings/change_settings.php:46]: Meal Plan

File: change_settings/change_settings.php [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+                    </li>
+                    <li class="nav-item">
+                        <a class="nav-link" href="../imprint/imprint.html">Support & Imprint</a>
+                    </li>
+                </ul>
+                <form class="form-inline my-2 my-lg-0 mr-sm-2" action="../login/login_form.php">
+                    <button class="btn btn-primary" type="submit">Log out</button>
```
Comment (Zylesto) [inline change_settings/change_settings.php:53]: Logout

File: change_settings/change_settings.php [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+                                                        <h6 class="mb">Email</h6>
+                                                    </div>
+                                                    <div class="col-md-9">
+                                                        <input id="email" type="text" class="form-control" value="<?php echo $_SESSION["email"] ?>" readonly=True>
+                                                    </div>
+                                                </div>
+                                                <hr>
+                                                <div class="row">
+                                                    <div class="col-md-3">
+                                                        <h6 class="mb">Birth day</h6>
```
Comment (Zylesto) [inline change_settings/change_settings.php:144]: Birthday

File: change_settings/change_settings.php [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+                                            </form>
+                                        </div>
+                                    </div>
+                                </div>
+                            </div>
+                            <div class="row">
+                                <div class="col"></div>
+                                <div class="col">
+                                    <div class="d-flex justify-content-center m-3">
+                                        <button class="btn btn-primary btn-sm" type="submit" onclick="changeAccountSettings()">Save changes</button>
```
Comment (Zylesto) [inline change_settings/change_settings.php:214]: btn btn-primary btn-lg + Save Changes

File: change_settings/change_settings.php [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+                                </div>
+                            </div>
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </div>
+
+
+<!--        </div>-->
```
Comment (Zylesto) [inline change_settings/change_settings.php:449]: was das

**Review by Zylesto — APPROVED** [review]

File: change_settings/change_settings_functionality.php (hunk 1 above)
Comment (Zylesto) [reply change_settings/change_settings_functionality.php:15]: Please also center the toasts and fix the navbar

**Comment by Zylesto** [comment]
Please also center the toasts and fix the navbar

File: change_settings/change_settings.php (hunk 7 above)
Comment (ordass00) [reply change_settings/change_settings.php:214]: sm is fine here

File: change_settings/change_settings.php (hunk 8 above)
Comment (ordass00) [reply change_settings/change_settings.php:449]: removed

File: change_settings/change_settings.php (hunk 7 above)
Comment (Zylesto) [reply change_settings/change_settings.php:214]: no large is better

File: change_settings/change_settings.ts [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+                showToastErrorMessage("error_toast_ingredients", "error_text_ingredients", data.errorText);
+            } else {
+                showToastMessage("successfully_changed_ingredients_toast");
+            }
+        })["catch"](function (error) {
+        showToastErrorMessage("error_toast_ingredients", "error_text_ingredients", error.errorText);
+    });
+}
+
+let pwd_input = (document.getElementById("new_password_input") as HTMLInputElement);
```
Comment (ArminShinobi) [inline change_settings/change_settings.ts:161]: What is this part of?

File: change_settings/change_settings.php [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+        window.changeIngredients = changeIngredients;
+        window.pwdValidation = pwdValidation;
+        window.showHidePassword = showHidePassword;
+        window.validateInput = validateInput;
+    </script>
+
+    <title>Account Settings</title>
+</head>
+<body>
+    <?php
```
Comment (ArminShinobi) [inline change_settings/change_settings.php:31]: We have put session_start at the top of every page

File: change_settings/change_settings_functionality.php [hunk 11]
```diff
@@ -0,0 +1,96 @@
+<?php
+session_start();
+include "../shared/php/database.php";
+
+header("Content-Type: application/json");
+$request = file_get_contents("php://input");
+$conn = connect_local();
+
+if (isset($request) && !empty($request)) {
```
Comment (ArminShinobi) [inline change_settings/change_settings_functionality.php:9]: Your else case is missing:
else {
    echo json_encode(["result" => null, "error" => true, "errorText" => "Request is not set or is empty."]);
}

File: change_settings/change_settings_functionality.php [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            } else {
+                echo json_encode(["error" => true, "errorText" => "Please fill out both input fields."]);
+            }
+            break;
+
+        case "change_preferences":
+            $intolerances_string = "";
+            $intolerances_array = $reqObj->intolerances;
+            if (empty($intolerances_array)) {
+                $intolerances_string = " ";
```
Comment (ArminShinobi) [inline change_settings/change_settings_functionality.php:57]: I think that should be an empty string not a string with blank in it

File: shared/php/database.php [hunk 13]
```diff
@@ -143,4 +155,42 @@ function clear_password_reset($conn, $user_id){
...
   $stmt->bindParam(1, $user_id);
   return $stmt->execute();
+}
+
+function change_username($conn, $user_id, $new_username)
+{
+  if ($conn == null || $user_id == null || $new_username == null) {
+    return null;
+  }
+  $sql = "UPDATE user SET Username=? WHERE ID=?;";
```
Comment (ArminShinobi) [inline shared/php/database.php:165]: Please make sure that this is the same database  table as in one.com

File: change_settings/change_settings_functionality.php (hunk 12 above)
Comment (ordass00) [reply change_settings/change_settings_functionality.php:57]: otherwise it doesnt work

File: shared/php/database.php (hunk 13 above)
Comment (ordass00) [reply shared/php/database.php:165]: in another pr

**Review by ArminShinobi — APPROVED** [review]

**Pull request merged**